game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java

```diff
@@ -340,43 +349,66 @@ void notifyRetreat(final Collection<Unit> retreating) {
     attackerModel.notifyRetreat(retreating);
   }
 
+  @Nullable
   Territory getRetreat(
       final String message, final Collection<Territory> possible, final boolean submerge) {
-    return (!submerge || possible.size() > 1)
-        ? getRetreatInternal(message, possible)
-        : getSubmerge(message);
-  }
-
-  private Territory getSubmerge(final String message) {
     if (SwingUtilities.isEventDispatchThread()) {
       throw new IllegalStateException("Should not be called from dispatch thread");
     }
-    final AtomicReference<Territory> retreatTo = new AtomicReference<>();
-    final CountDownLatch latch = new CountDownLatch(1);
-    final Action action =
-        SwingAction.of(
-            "Submerge Subs?",
-            e -> {
-              actionButton.setEnabled(false);
-              if (showSubmergeDialog(message, retreatTo)) {
-                actionButton.setAction(nullAction);
-                latch.countDown();
-              }
-              actionButton.setEnabled(true);
-            });
+    final CompletableFuture<Territory> future = new CompletableFuture<>();
+    final String title;
+    final Supplier<RetreatResult> supplier;
+    if (!submerge || possible.size() > 1) {
+      title = "Retreat?";
+      supplier = () -> showRetreatDialog(message, possible);
+    } else {
+      title = "Submerge Subs?";
+      supplier = () -> showSubmergeDialog(message);
+    }
+    final Action action = getPlayerAction(title, supplier, future);
     SwingUtilities.invokeLater(
         () -> {
           actionButton.setAction(action);
           action.actionPerformed(null);
         });
-    mapPanel.getUiContext().addShutdownLatch(latch);
-    Interruptibles.await(latch);
-    mapPanel.getUiContext().removeShutdownLatch(latch);
-    return retreatTo.get();
+
+    return awaitUserInput(future);
+  }
+
+  private Territory awaitUserInput(final CompletableFuture<Territory> future) {
+    final Active rejectionCallback =
+        () -> future.completeExceptionally(new RuntimeException("Shutting down"));
+    try {
+      mapPanel.getUiContext().addActive(rejectionCallback);
+      return future.get();
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    } catch (final ExecutionException e) {
+      log.log(Level.INFO, "UiContext shut down before supplying result", e);
+    } finally {
+      mapPanel.getUiContext().removeActive(rejectionCallback);
+    }
+    return null;
+  }
```

Comment:
Seems like this should live somewhere else and be generic to the return type. Since it's using UiContext, perhaps we can add the method on UiContext instead?

Reply:
Valid point, but I saw this kind of pattern in a lot of places, so that would be a larger cleanup.
I plan renaming `removeActive` and `addActive` to `addShutdownHook` and `removeShutdownHook` and replace the `Active` interface with runnable in a follow-up PR.
I think the current status is definitely an improvement for now, but I agree that it can be improved further.
I'd move this method as part of this incoming PR then.
Any further refactoring is not on my priority list though, so it would be up to you in case you want to apply this pattern elsewhere.